### PR TITLE
cp pointers util from kubernetes/kubernetes

### PR DESCRIPTION
**Pull request by stewart-yu**
We need move the` k8s.io/kubernetes/pkg/util/pointer` package to` k8s.io/utils/` repo, more details please refer to [PR](https://github.com/kubernetes/kubernetes/pull/66284) in `kubernetes` repo.

Following the [instruction](https://github.com/kubernetes/utils/blob/master/HOWTOMOVE.md), this PR is the first step

### CODE REVIEW

**Comment by stewart-yu** [comment]
seems no `OWNER` file in this repo
/assign @thockin @luxas @sttts 
PTAL, related kubernetes repo [PR](https://github.com/kubernetes/kubernetes/pull/66284), thanks

File: README.md [hunk 1]
```diff
@@ -1,5 +1,6 @@
 # Utils
 
+
```
Comment (sttts) [inline README.md]: nit: why this empty line?

File: pointer/pointer.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+	if ptr != nil {
+		return *ptr
+	}
+	return def
+}
+
+// BoolPtr returns a pointer to a bool
+func BoolPtr(b bool) *bool {
+	return &b
+}
```
Comment (sttts) [inline pointer/pointer.go]: StringPtr. Don't we have that 1000 times in kube?

Comment (stewart-yu) [reply pointer/pointer.go]: sure,  add it now

**Comment by stewart-yu** [comment]
/cc  @thockin 
for review, thanks

File: README.md [hunk 3]
```diff
@@ -45,6 +45,7 @@ an existing package to this repository.
 
 - [Clock](/clock) provides an interface for time-based operations.  It allows
   mocking time for testing.
+- [Pointers](/pointers) provides some functions for pointer-based operations.
```
Comment (apelisse) [inline README.md]: For some reason, we've kept an empty line between each module (don't ask me why), but maybe we should keep it consistent?

File: pointer/pointer.go [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+	}
+	for i := 0; i < v.NumField(); i++ {
+		if v.Field(i).Kind() == reflect.Ptr && !v.Field(i).IsNil() {
+			return false
+		}
+	}
+	return true
+}
+
+// Int32Ptr returns a pointer to an int32
```
Comment (apelisse) [inline pointer/pointer.go]: AFAICR, Tim gave this as an example of what we shouldn't have in this repo (might remember it wrong, I'll look it up)

**Comment by apelisse** [comment]
https://github.com/kubernetes/kubernetes/issues/48209#issuecomment-312305457

File: pointer/pointer.go (hunk 4 above)
Comment (stewart-yu) [reply pointer/pointer.go]: seems `Int32Ptr()` had been used, if do not put here,  which file fo put it?

Comment (apelisse) [reply pointer/pointer.go]: The problem is not that it's not being used, the problem is that you're trying to share code with very little complexity.

Comment (luxas) [reply pointer/pointer.go]: It's gonna be used by many packages, and is generally useful, although very simple.

**Comment by luxas** [comment]
On the other hand @apelisse https://github.com/kubernetes/kubernetes/pull/66010#pullrequestreview-135948977 :)

> This seems fine but why apimachinery and not https://github.com/kubernetes/utils ? I might clean up some of the APIs, if you want to put a little polish into it...

**Comment by apelisse** [comment]
I guess I can't argue with that :-)

File: pointer/pointer_test.go [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+			Bar *int
+		}{nil, new(int)}, false},
+		{&struct {
+			Foo *int
+			Bar *int
+		}{nil, new(int)}, false},
+		{(*struct{})(nil), true},
+	}
+	for i, tc := range testCases {
+		if AllPtrFieldsNil(tc.obj) != tc.expected {
```
Comment (apelisse) [inline pointer/pointer_test.go]: This could use `t.Run`

**Comment by luxas** [comment]
Can you merge then to unblock us @apelisse :)?

**Comment by luxas** [comment]
Thanks :tada:!

**Comment by stewart-yu** [comment]
Thanks all

**Comment by sttts** [comment]
Finally we have a generic pointer package 🎉🚀

File: pointer/pointer.go [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+	return def
+}
+
+// BoolPtr returns a pointer to a bool
+func BoolPtr(b bool) *bool {
+	return &b
+}
+
+// StringPtr returns a pointer to the passed string.
+func StringPtr(s string) *string {
```
Comment (sttts) [inline pointer/pointer.go]: float64 is missing 😱

**Comment by thockin** [comment]
This is in so I won't be a pain about it, but this is the sort of dependency that causes very interconnected dep graphs.  Any tool that uses this now has to vendor this whole repo.  Part of the go ethos is that it's better to copy a little code than to take a dependency.  Let's please not let this accumulate tons of random crap.

**Comment by luxas** [comment]
No agree we shouldn't put "random crap" in this repo, but this exact utilpointer code is referenced by most of our API packages that deal with defaulting, so this is kinda-needed unless we want to start doing stuff like this in our defaulting packages (that will now live in `n` different repos)

```go
trueVar := true
obj.SomeBoolPointerField = &trueVar
```
vs

```go
obj.SomeBoolPointerField = utilpointer.BoolVar(true)
```

Well you know this already, but anyway I've made the argument for this now.

For me this could have moved to `k8s.io/apimachinery` equally fine but you said in https://github.com/kubernetes/kubernetes/pull/66010#pullrequestreview-135948977

> This seems fine but why apimachinery and not https://github.com/kubernetes/utils ? I might clean up some of the APIs, if you want to put a little polish into it...

So I'm a bit confused.
In any case I'm of the opinion this needs to live somewhere to not be a pain for the defaulting code for the ComponentConfigs.

> Any tool that uses this now has to vendor this whole repo

hmm I had the idea one can vendor individual packages, not the whole repo? @sttts had also done some experimenting with this I think wrt the ComponentConfig KEP in general

**Comment by thockin** [comment]
Lucas, you should know by now that I never claim to be 100% self-consistent
:)

You can vendor just one package, but you have to download and examine the
whole repo.  Some tools used to not do subdir-vendoring, but I think we're
past that.

**Comment by luxas** [comment]
> Lucas, you should know by now that I never claim to be 100% self-consistent

Hahah, yeah ;). Can't say I'm always super consistent with reviews either lol

> You can vendor just one package, but you have to download and examine the
whole repo.  Some tools used to not do subdir-vendoring, but I think we're
past that.

I don't think this will be a problem as anything should be able to depend on `k8s.io/utils`, this is used by multiple k8s consumers in various repos, the package has a well-defined scope, and all modern vendoring systems I know of supports pulling individual packages.

@thockin If this sounds good to you, go ahead and approve https://github.com/kubernetes/kubernetes/pull/66284 which follows this up in the k/k repo ;)

**Pull request merged**